### PR TITLE
fix for unversioned requests without action

### PR DESCRIPTION
**Pull request by wjdan94**
- Handles only un-versioned requests to /volume and /image with no action. 
- #FIXME: need to update "updated" date and time if OS makes any changes to volume API.
- #TODO: move the new method _list_api_versions() to services.py, better?
- for _list_api_versions(), instead of hardcoding a template/example for the API list and update it, I recreated the list from scratch instead. Better or worse?

### CODE REVIEW

File: mixmatch/proxy.py [hunk 1]
```diff
@@ -169,13 +195,13 @@ def _search_forward(self):
         for sp in CONF.proxy.service_providers:
             if sp == 'default':
                 response = self._do_request_on('default')
-                if 200 <= response.status_code < 300:
+                if 200 <= response.status_code <= 300:
```
Comment (gsilvis) [inline mixmatch/proxy.py]: Why is this being changed?

Comment (wjdan94) [inline mixmatch/proxy.py]: that is needed *only if* want to forward the call with no action to sp's, and for volume api, calls with no action return 300. BUT we are not following this approach anymore, so this has been fixed.

Comment (gsilvis) [inline mixmatch/proxy.py]: Makes sense

File: mixmatch/proxy.py [hunk 2]
```diff
@@ -205,6 +217,66 @@ def _aggregate_forward(self):
...
+                    info['min_version'] = version[1:]
+                    info['version'] = version[1:]
+                else:
+                    info['status'] = 'SUPPORTED'
+                    info['min_version'] = ''
+                    info['version'] = ''
+
+                info['id'] = version
+                info['updated'] = '2014-06-28T12:20:21Z'  # FIXME
+                info['links'] = list()
```
Comment (wjdan94) [inline mixmatch/proxy.py]: Done!

File: mixmatch/config.py [hunk 3]
```diff
@@ -50,7 +50,15 @@
 
     cfg.IntOpt('cache_time',
                default=600,
-               help='How long to store cached tokens for')
+               help='How long to store cached tokens for'),
+
+    cfg.ListOpt('image_api_versions',
+                default=[],
```
Comment (gsilvis) [inline mixmatch/config.py]: Probably supply a reasonable default for this [all of them?]

File: mixmatch/config.py [hunk 4]
```diff
@@ -50,7 +50,15 @@
...
+                help='List of supported image api versions'),
+
+    cfg.ListOpt('volume_api_versions',
+                default=[],
```
Comment (gsilvis) [inline mixmatch/config.py]: Ditto

File: mixmatch/proxy.py [hunk 5]
```diff
@@ -53,60 +54,74 @@ def __init__(self, method, path, headers):
...
+                # Which we already know the location of, use that SP.
+                self.service_provider = mapping.resource_sp
+                self.project_id = mapping.tenant_id
+                self._forward = self._targeted_forward
+            else:
+                self._forward = self._search_forward
+
+        elif len(self.request_path) < 2:
+            # calls with no version and action
+            self._forward = self._list_api_versions
```
Comment (gsilvis) [inline mixmatch/proxy.py]: Can you put this part first?  That way, you can put a return in the ``if`` block and avoid indenting everything that comes after it

Comment (gsilvis) [inline mixmatch/proxy.py]: As a bonus, this'll make the diff a lot smaller :)

**Comment by knikolla** [comment]
Hmmm... Also can you please write some unit tests for this?

File: mixmatch/proxy.py [hunk 6]
```diff
@@ -53,7 +54,17 @@ def __init__(self, method, path, headers):
             self.request_path.insert(0, 'image')
 
         self.service_type = self.request_path[0]
+
+        if len(self.request_path) == 1:
+            # unversioned calls with no action
+            self._forward = self._list_api_versions
+            return None
```
Comment (gsilvis) [inline mixmatch/proxy.py]: I'd suggest just 'return' here---we're not returning a value anywhere else

**Pull request merged**